operator-common/src/main/java/io/strimzi/operator/common/AbstractOperator.java

```diff
@@ -40,20 +51,22 @@
  *
  * <ul>
  * <li>uses the Fabric8 kubernetes API and implements
- * {@link #reconcile(Reconciliation)} by delegating to abstract {@link #createOrUpdate(Reconciliation, HasMetadata)}
+ * {@link #reconcile(Reconciliation)} by delegating to abstract {@link #createOrUpdate(Reconciliation, AbstractCustomResource)}
  * and {@link #delete(Reconciliation)} methods for subclasses to implement.
  * 
- * <li>add support for operator-side {@linkplain #validate(HasMetadata) validation}.
+ * <li>add support for operator-side {@linkplain #validate(AbstractCustomResource) validation}.
  *     This can be used to automatically log warnings about source resources which used deprecated part of the CR API.
  *
  * </ul>
  * @param <T> The Java representation of the Kubernetes resource, e.g. {@code Kafka} or {@code KafkaConnect}
- * @param <S> The "Resource Operator" for the source resource type. Typically this will be some instantiation of
+ * @param <O> The "Resource Operator" for the source resource type. Typically this will be some instantiation of
  *           {@link io.strimzi.operator.common.operator.resource.CrdOperator}.
  */
 public abstract class AbstractOperator<
-        T extends HasMetadata,
-        S extends AbstractWatchableResourceOperator<?, T, ?, ?, ?>>
+        T extends AbstractCustomResource<P, S>,
+        P extends Spec,
+        S extends Status,
+        O extends AbstractWatchableResourceOperatorWithStatus<?, T, ?, ?, ?>>
```

Comment:
It shouldn't be in this PR, but I think we should consider using proper names rather than single letters for type parameters, since it remembering the single letters gets harder the more thee are of them.